### ISSUE TITLE
Cant even get it started!

### ISSUE THREAD

**Issue by gfmoore** [CommaSword/node-play-mp3#1]
Hi,

I am hoping this project succeeds as I really would like a node.js version of audio to work :)

However, I can't even get the sample code you provide to work or be interpreted.

System I'm using is Windows 10 (I'm guessing this may be an issue as it is with most other players for node.

I've installed Node v v8.9.4
I've installed electron 1.8.2
i've installed you node-mp3-player

added: I've also installed the dependencies I've noted as well, JSONStream, electron-spawn, uuid

Now I'm not too sure about electron and there isn't any particular direction on your page, but I required it anyway?

I pasted in your sample code into a test.js file and using vs code debug ran it.  (also tried using node test.js)

get following error

Debugging with inspector protocol because Node.js v8.9.4 was detected.
node --inspect-brk=44492 test.js 
Debugger listening on ws://127.0.0.1:44492/5e059de3-ce31-4512-86ad-0aca6b82cd47
Debugger attached.
c:\GMMusicServer\test.js:8
})();
  ^

TypeError: createAudio(...)(...) is not a function
    at Object.<anonymous> (c:\GMMusicServer\test.js:8:3)
    at Module._compile (module.js:640:14)
    at Object.Module._extensions..js (module.js:654:10)
    at Module.load (module.js:556:32)
    at tryModuleLoad (module.js:499:12)
    at Function.Module._load (module.js:491:3)
    at Function.Module.runMain (module.js:684:10)
    at startup (bootstrap_node.js:187:16)
    at bootstrap_node.js:608:3
Waiting for the debugger to disconnect...
TypeError: createAudio(...)(...) is not a function
    at Object.<anonymous> (c:\GMMusicServer\test.js:8:3)
    at Module._compile (module.js:640:14)
    at Object.Module._extensions..js (module.js:654:10)
    at Module.load (module.js:556:32)
    at tryModuleLoad (module.js:499:12)
    at Function.Module._load (module.js:491:3)
    at Function.Module.runMain (module.js:684:10)
    at startup (bootstrap_node.js:187:16)
    at bootstrap_node.js:608:3

Okay, first I didn't realise you didn't need semi colons anymore :)
This code looks like an immediately invoked function expression so I can't see why there is an issue here.
I haven't come across async and await before - doing some research now.

I'm suspecting it's because i am  on windows or because you are expecting me to run this with electron?? If so how?

Thanks for any pointers you can provide :)

Here's hoping.

regards

Gordon Moore (no not that one...)

**Comment by imsnif** [comment]
Hey Gordon!

Indeed I did not get a chance to test this on windows, but I'm hoping we can get it to work anyway.
Just to remove some elements from the equation, could you run it outside vscode? Just by pasting the example code as is in your `test.js` file and running `node test.js`? Do you get the same error?

**Comment by gfmoore** [comment]
Heyup

Yes I got the same error running node test.js


I've tried the following code (adding in the semi-colons and a try catch block

const electron = require('electron');

const { createAudio } = require('node-mp3-player');
const Audio = createAudio();

(async () => { 
  try {
    const myFile = await Audio(`${__dirname}/test1.mp3`);
    await myfile.play();
  }
  catch (e) {
    console.log('error ' + e.toString());
  }
})();


I am now getting an error 

error ReferenceError: myfile is not defined

seems like it's not finding my mp3 file?? It's in the same folder as test.js

I am very happy to help here, but I am a little newbi ish on some points lol

**Comment by imsnif** [comment]
Okay @gfmoore - I think I found the issue. It was actually my mistake in the example code. One of the only places in which you DO need a semicolon is before an iife. :)
I updated the `README` - want so see if this works out for you?

And btw - you don't need to install `electron` or any such thing. This library creates its instance on its own behind the scenes using `electron-spawn`.

P.S.
In your pasted example the issue is that in the `await myfile.play();` it should by `myFile`
Also - very glad for your help! Thank you for it.

**Comment by gfmoore** [comment]
Just for completeness - your code exactly

λ node test1
C:\GMMusicServer\test1.js:10
})()
  ^

TypeError: createAudio(...)(...) is not a function
    at Object.<anonymous> (C:\GMMusicServer\test1.js:10:3)
    at Module._compile (module.js:643:30)
    at Object.Module._extensions..js (module.js:654:10)
    at Module.load (module.js:556:32)
    at tryModuleLoad (module.js:499:12)
    at Function.Module._load (module.js:491:3)
    at Function.Module.runMain (module.js:684:10)
    at startup (bootstrap_node.js:187:16)
    at bootstrap_node.js:608:3

**Comment by gfmoore** [comment]
Okay - can't believe the typo... doh schoolboy error not to check!

(and as for the one semi ... I even looked this up and some guru said don't use them ... :)

Anyhow still not hearing anything

const { createAudio } = require('node-mp3-player')
const Audio = createAudio();
 
(async () => {

  try {
  const myFile = await Audio(`${__dirname}/test.mp3`)
  await myFile.play()
  }
  catch (e) {
    console.log('error: ' + e.toString())
  }
})()

λ node test.js
error: Error: timed out waiting for electron process

**Comment by gfmoore** [comment]
const myFile = await Audio(${'test.mp3')  

still no joy. Just noted that the back ticks don't paste here!

I know the javascript audio object works in a normal html/javascript file.

I'm guessing it's not finding a device to output to?

**Comment by gfmoore** [comment]
Oh and I'm putting in a try catch block because of the warning:


(node:16748) UnhandledPromiseRejectionWarning: Unhandled promise rejection (rejection id: 1): Error: timed out waiting for electron process
warning.js:18
(node:16748) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.

I'm sure there's a better pattern than this, but I've just started to read up on it.

**Comment by imsnif** [comment]
Hey no worries - as long as you can see the error.
Okay, this might be a platform issue or a bug. Could you try to kill any process called `electron` on the computer? Maybe there are some leftovers from before.

**Comment by imsnif** [comment]
(and just to be sure - could you give the Audio constructor an absolute path? This shouldn't be needed, but it will be easier to rule it out... like: `C:\path\to\my\mp3\file.mp3` - just as a normally quoted string)

**Comment by gfmoore** [comment]
Can't see any process linked to electron - though being windows that doesn't mean anything...(you can't even search for it afaik)

I've tried quite a few node packages for this and they all seem to have the same problem - no output. I'm sure it's linked to not finding a device to send the info to. Unfortunately none of them seem to allow you to declare a device.

Will try the absolute path.

**Comment by gfmoore** [comment]
sorry no joy, no matter what combo of forward or backward slashes or double back slashes!

Still get same message.

**Comment by imsnif** [comment]
This library creates an `electron` child process and communicates with it in order to use the DOM Audio API it provides. This error means there is a problem reaching that child process. This might indeed be a platform issue (as you say, this was historically a thing, I'm not sure what's going on with it now: https://github.com/electron/electron/issues/964), but it might just mean the timeout needs adjustment.
So if you'll bare with me for one more try, could you edit `node_modules/node-mp3-player/index.js` and in line 31 change `500` to `5000`? This will increase the timeout to 5 seconds which should be more than enough. I apologize that there is no way to configure this properly yet. If we find that this is the issue, I'll patch it up.

**Comment by gfmoore** [comment]
mate, there is no problem with me trying as much as you need. If I can help get it working then my gain.

on it.

**Comment by gfmoore** [comment]
No joy I'm afraid.

It waited the 5 seconds and same error. 

I've tried inspecting the myFile object (with a breakpoint) but can't see anything untoward, but certainly there is no evidence of the test1.mp3 filename either? Should there be?

**Comment by gfmoore** [comment]
I have just tested the test1.mp3 - it's only a few seconds of Free - I'm alright now ( :) )  and that works fine. On its own, not from the app I hasten to add.

**Comment by imsnif** [comment]
Sorry, could you please explain your last? I didn't understand.

**Comment by gfmoore** [comment]
Sorry, I was saying that I manually tested the test1.mp3 file (from the command line) to ensure that it wasn't corrupted. It plays fine using vlc.

The title of the song is Free - I'm alright now  - just being a bit sardonic (English humour)

**Comment by imsnif** [comment]
No worries. :)
So is probably some issue `electron` has with your sound card. Which makes sense considering you said other libraries had the same issue as well. Could be an edge case (more than one sound card in windows) that commonly gets ignored.
I don't know a lot about windows - my apologies for not being able to offer much helpful advice.

**Comment by gfmoore** [comment]
Hi thanks for all your efforts

FYI: I have managed to get node-mpg123-util to output some sound - hurrah, so it is possible.
https://www.npmjs.com/package/node-mpg123-util

Added: This is not as useful as I hoped. It will play an mp3 and change the volume, but it does not offer a  pause option that I can figure. It seems to stream the mp3 and encode it to pcm using Lame encoder/decoder. It is a start though, but doesn't use the nice Audio object that this project uses. Pity

Hope you don't mind me referencing an alternative for windows users, but if anyone else comes here at least there is hope :)

(They may? need to install speaker, which needs python installing:
npm --add-python-to-path='true' --debug install --global windows-build-tools (from an elevated powershell prompt)
then npm install speaker.

Hopefully one day you or someone might find the solution as what you are doing looks good.

Gordon
